### ISSUE TITLE
fix the  (_rconcat #0 #0) invention cost mismatch

### ISSUE THREAD

**Issue by gabegrand** [mlb2251/stitch#129]
err msg:
```
Stats { worklist_steps: 718, finished: 3, calc_final_utility: 122, upper_bound_fired: 15, free_vars_fired: 103, single_use_fired: 307, single_task_fired: 289, useless_abstract_fired: 69, force_multiuse_fired: 0 }
Cost before: 22021
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `20607`,
 right: `20506`: 
(_rconcat #0 #0): utility_upper_bound=2928, body_utility=(102,0), refinements=0, match_locations=3, usages=10 -> finished: utility=1413, compressive_utility=1515, arity=1, usages=10
', src/extraction.rs:157:9
```

cmd:
```
cd stitch; cargo run --bin=compress --release -- /Users/grandg/mit/laps/experiments_iterative/outputs/gg_laps_domains/domains/re2/stitch/seed_111/stitch_10/0/train/stitch_frontiers.json --out=/Users/grandg/mit/laps/experiments_iterative/outputs/gg_laps_domains/domains/re2/stitch/seed_111/stitch_10/0/train/stitch_output.json --max-arity=2 --iterations=10 --fmt=dreamcoder
```

**Comment by mlb2251** [comment]
Adding a quick writeup from what I found a few days ago. TLDR it's a pretty rare thing that is a utility function issue and the rewritten result is actually correct it's just the utility function that's wrong. Was too sick to properly address it when I found it but can give another shot

Say we have the invention `inv = foo #0 #0`

then we try to compress `(foo (foo a a)  (foo a a))`

this should yield `(inv (inv a))` and it correctly does, however there will be a cost mismatch where our current utility function gets this slightly wrong.

This is because our utility sees 3 places we can rewrite in `(foo (foo a a)  (foo a a))`, however it doesnt account for the fact that multiuse on the outer rewrite location will effectively mean you only get rewrite gains from one of the two `(foo a a)`

**Comment by mlb2251** [comment]
This was also observed with `[fn_4 arity=2: (#1 #0 #0)]` see slack

**Comment by mlb2251** [comment]
`--utility-by-rewrite` is now a better workaround than `--no-cost-mismatch` for this, however it increases runtime by around 10x

**Comment by mlb2251** [comment]
fixed in 823af6976b8f !
